plan/logical_plan_builder.go

```diff
@@ -1709,6 +1709,10 @@ func (b *planBuilder) buildDataSource(tn *ast.TableName) LogicalPlan {
 	tableInfo := tbl.Meta()
 	b.visitInfo = appendVisitInfo(b.visitInfo, mysql.SelectPriv, dbName.L, tableInfo.Name.L, "")
 
+	if tableInfo.GetPartitionInfo() != nil {
+		b.optFlag = b.optFlag | flagTablePartition
```

Comment:
We should clear the flag before leaving buildDataSource.

Reply:
We new a planBuilder instance everytime in the `Optimize` function, the object would not be reused, so the flag need not to be clear.